File: frontend/extract_data.js
/*
  extract_data.js — Run with: node extract_data.js
  
  Reads DATA/Initial Dataset/CSV/Data.csv (semicolon-separated, comma decimals)
  and produces:
    - frontend/data/translators.json
    - frontend/data/clients.json
*/
const fs = require('fs');
const path = require('path');

const CSV_PATH = path.join(__dirname, '..', 'DATA', 'Initial Dataset', 'CSV', 'Data.csv');
const OUT_DIR  = path.join(__dirname, 'data');
if (!fs.existsSync(OUT_DIR)) fs.mkdirSync(OUT_DIR, { recursive: true });

console.log('Reading CSV...');
const raw = fs.readFileSync(CSV_PATH, 'utf-8');
const lines = raw.split('\n').filter(l => l.trim());
const headers = lines[0].split(';').map(h => h.trim());
console.log(`  ${lines.length - 1} rows, ${headers.length} cols`);

// Parse rows
const rows = [];
for (let i = 1; i < lines.length; i++) {
  const vals = lines[i].split(';');
  if (vals.length < headers.length) continue;
  const obj = {};
  headers.forEach((h, j) => { obj[h] = (vals[j] || '').trim(); });
  rows.push(obj);
}
console.log(`  Parsed ${rows.length} rows.`);

function commaFloat(s) { return parseFloat((s || '0').replace(',', '.')); }

// ── TRANSLATORS ──────────────────────────────────────────────
console.log('\nExtracting translators...');
const trMap = {};
for (const r of rows) {
  const name = r.TRANSLATOR;
  if (!name) continue;
  if (!trMap[name]) trMap[name] = { tasks: new Set(), quality: [], rate: [], hours: 0, types: new Set(), srcs: new Set(), tgts: new Set(), clients: new Set(), pairs: {} };
  const t = trMap[name];
  t.tasks.add(r.TASK_ID);
  if (r.QUALITY_EVALUATION) t.quality.push(commaFloat(r.QUALITY_EVALUATION));
  if (r.HOURLY_RATE) t.rate.push(commaFloat(r.HOURLY_RATE));
  t.hours += commaFloat(r.HOURS);
  if (r.TASK_TYPE) t.types.add(r.TASK_TYPE);
  if (r.SOURCE_LANG) t.srcs.add(r.SOURCE_LANG);
  if (r.TARGET_LANG) t.tgts.add(r.TARGET_LANG);
  if (r.MANUFACTURER) t.clients.add(r.MANUFACTURER);
  const pairKey = `${r.SOURCE_LANG}→${r.TARGET_LANG}`;
  t.pairs[pairKey] = (t.pairs[pairKey] || 0) + 1;
}

const translators = Object.entries(trMap).map(([name, t], i) => {
  const avg = arr => arr.length ? Math.round(arr.reduce((a,b) => a+b, 0) / arr.length * 10) / 10 : 0;
  // Find most common pair
  let topPair = { src: 'N/A', tgt: 'N/A' };
  let topCount = 0;
  for (const [pk, cnt] of Object.entries(t.pairs)) {
    if (cnt > topCount) { topCount = cnt; const [s, tg] = pk.split('→'); topPair = { src: s, tgt: tg }; }
  }
  return {
    id: `TR-${String(i + 1).padStart(4, '0')}`,
    name,
    source: topPair.src,
    target: topPair.tgt,
    rate: avg(t.rate),
    quality: avg(t.quality),
    taskTypes: [...t.types],
    taskCount: t.tasks.size,
    totalHours: Math.round(t.hours * 10) / 10,
    sourceLangs: [...t.srcs],
    targetLangs: [...t.tgts],
    clientsWorked: t.clients.size,
    status: 'Available',
    workerType: t.tasks.size > 20 ? 'Internal' : 'Third-Party',
  };
});

fs.writeFileSync(path.join(OUT_DIR, 'translators.json'), JSON.stringify(translators));
console.log(`  ${translators.length} translators → data/translators.json`);

// ── CLIENTS ──────────────────────────────────────────────────
console.log('\nExtracting clients...');
const clMap = {};
for (const r of rows) {
  const name = r.MANUFACTURER;
  if (!name) continue;
  if (!clMap[name]) clMap[name] = { industry: r.MANUFACTURER_INDUSTRY, sector: r.MANUFACTURER_SECTOR, industryGroup: r.MANUFACTURER_INDUSTRY_GROUP, subindustry: r.MANUFACTURER_SUBINDUSTRY, tasks: new Set(), quality: [], rate: [], translators: new Set(), types: new Set() };
  const c = clMap[name];
  c.tasks.add(r.TASK_ID);
  if (r.QUALITY_EVALUATION) c.quality.push(commaFloat(r.QUALITY_EVALUATION));
  if (r.HOURLY_RATE) c.rate.push(commaFloat(r.HOURLY_RATE));
  if (r.TRANSLATOR) c.translators.add(r.TRANSLATOR);
  if (r.TASK_TYPE) c.types.add(r.TASK_TYPE);
}

const clients = Object.entries(clMap).map(([name, c], i) => {
  const avg = arr => arr.length ? Math.round(arr.reduce((a,b) => a+b, 0) / arr.length * 10) / 10 : 0;
  return {
    id: `CL-${String(i + 1).padStart(4, '0')}`,
    name,
    industry: c.industry || 'General',
    sector: c.sector || '',
    industryGroup: c.industryGroup || '',
    subindustry: c.subindustry || '',
    taskCount: c.tasks.size,
    avgRate: avg(c.rate),
    avgQuality: avg(c.quality),
    translatorsUsed: c.translators.size,
    taskTypes: [...c.types],
  };
});

fs.writeFileSync(path.join(OUT_DIR, 'clients.json'), JSON.stringify(clients));
console.log(`  ${clients.length} clients → data/clients.json`);
console.log('\nDone!');

File: frontend/extract_data.py
"""
extract_data.py — Generate JSON data files for the IDISC HUD frontend.
Reads Data.csv (semicolon-separated, comma decimal) and produces:
  - translators.json  (~900 unique translators with aggregated stats)
  - clients.json      (unique clients/manufacturers with industry info)
"""
import pandas as pd
import json
import os

SCRIPT_DIR = os.path.dirname(os.path.abspath(__file__))
DATA_CSV = os.path.join(SCRIPT_DIR, "..", "DATA", "Initial Dataset", "CSV", "Data.csv")
OUT_DIR = os.path.join(SCRIPT_DIR, "data")
os.makedirs(OUT_DIR, exist_ok=True)

print(f"Reading {DATA_CSV} ...")
df = pd.read_csv(DATA_CSV, sep=";", decimal=",", encoding="utf-8", low_memory=False)
print(f"  {len(df):,} rows loaded.")

# ── TRANSLATORS ──────────────────────────────────────────────────────────────
print("\nExtracting translators ...")
# Aggregate per translator
tg = df.groupby("TRANSLATOR").agg(
    task_count=("TASK_ID", "nunique"),
    avg_quality=("QUALITY_EVALUATION", "mean"),
    avg_rate=("HOURLY_RATE", "mean"),
    total_hours=("HOURS", "sum"),
    task_types=("TASK_TYPE", lambda x: list(x.dropna().unique())),
    source_langs=("SOURCE_LANG", lambda x: list(x.dropna().unique())),
    target_langs=("TARGET_LANG", lambda x: list(x.dropna().unique())),
    clients_worked=("MANUFACTURER", lambda x: list(x.dropna().unique())),
    last_task=("START", "max"),
).reset_index()

translators = []
for _, r in tg.iterrows():
    # Determine primary language pair (most common)
    tr_data = df[df["TRANSLATOR"] == r["TRANSLATOR"]]
    pair_counts = tr_data.groupby(["SOURCE_LANG", "TARGET_LANG"]).size().reset_index(name="count")
    top_pair = pair_counts.sort_values("count", ascending=False).iloc[0] if len(pair_counts) > 0 else None

    translators.append({
        "id": f"TR-{len(translators)+1:04d}",
        "name": r["TRANSLATOR"],
        "source": top_pair["SOURCE_LANG"] if top_pair is not None else "N/A",
        "target": top_pair["TARGET_LANG"] if top_pair is not None else "N/A",
        "rate": round(r["avg_rate"], 1) if pd.notna(r["avg_rate"]) else 0,
        "quality": round(r["avg_quality"], 1) if pd.notna(r["avg_quality"]) else 0,
        "taskTypes": r["task_types"],
        "taskCount": int(r["task_count"]),
        "totalHours": round(r["total_hours"], 1) if pd.notna(r["total_hours"]) else 0,
        "sourceLangs": r["source_langs"],
        "targetLangs": r["target_langs"],
        "clientsWorked": len(r["clients_worked"]),
        "status": "Available",
        "workerType": "Internal" if r["task_count"] > 20 else "Third-Party",
    })

out_path = os.path.join(OUT_DIR, "translators.json")
with open(out_path, "w", encoding="utf-8") as f:
    json.dump(translators, f, ensure_ascii=False, indent=None)
print(f"  {len(translators)} translators → {out_path}")

# ── CLIENTS ──────────────────────────────────────────────────────────────────
print("\nExtracting clients ...")
cg = df.groupby("MANUFACTURER").agg(
    industry=("MANUFACTURER_INDUSTRY", "first"),
    sector=("MANUFACTURER_SECTOR", "first"),
    industry_group=("MANUFACTURER_INDUSTRY_GROUP", "first"),
    subindustry=("MANUFACTURER_SUBINDUSTRY", "first"),
    task_count=("TASK_ID", "nunique"),
    avg_sell_rate=("HOURLY_RATE", "mean"),
    avg_quality=("QUALITY_EVALUATION", "mean"),
    translators_used=("TRANSLATOR", "nunique"),
    task_types=("TASK_TYPE", lambda x: list(x.dropna().unique())),
).reset_index()

clients = []
for _, r in cg.iterrows():
    clients.append({
        "id": f"CL-{len(clients)+1:04d}",
        "name": r["MANUFACTURER"],
        "industry": r["industry"] if pd.notna(r["industry"]) else "General",
        "sector": r["sector"] if pd.notna(r["sector"]) else "",
        "industryGroup": r["industry_group"] if pd.notna(r["industry_group"]) else "",
        "subindustry": r["subindustry"] if pd.notna(r["subindustry"]) else "",
        "taskCount": int(r["task_count"]),
        "avgRate": round(r["avg_sell_rate"], 1) if pd.notna(r["avg_sell_rate"]) else 0,
        "avgQuality": round(r["avg_quality"], 1) if pd.notna(r["avg_quality"]) else 0,
        "translatorsUsed": int(r["translators_used"]),
        "taskTypes": r["task_types"],
    })

out_path = os.path.join(OUT_DIR, "clients.json")
with open(out_path, "w", encoding="utf-8") as f:
    json.dump(clients, f, ensure_ascii=False, indent=None)
print(f"  {len(clients)} clients → {out_path}")

print("\nDone!")
